Recast/Source/RecastArea.cpp

```diff
@@ -16,573 +16,700 @@
+        // This isn't normalized because it's defining the distance and direction the corner will need to be
+        // adjusted proportionally to the edge offsets to properly miter the adjoining edges.
+		float cornerMiterX = (prevSegmentNormX + currSegmentNormX) * 0.5f;
+		float cornerMiterZ = (prevSegmentNormZ + currSegmentNormZ) * 0.5f;
+        const float cornerMiterSqMag = rcSqr(cornerMiterX) + rcSqr(cornerMiterZ);
+
+        // If the magnitude of the segment normal average is less than about .69444,
+        // the corner is an acute enough angle that the result should be beveled.
+        const bool bevel = cornerMiterSqMag * MITER_LIMIT * MITER_LIMIT < 1.0f;
+
+        // Scale the corner miter so it's proportional to how much the corner should be offset compared to the edges.
+		if (cornerMiterSqMag > EPSILON)
 		{
-			const float scale = 1.0f / dmr2;
-			dmx *= scale;
-			dmy *= scale;
+			const float scale = 1.0f / cornerMiterSqMag;
+            cornerMiterX *= scale;
+            cornerMiterZ *= scale;
 		}
 
-		if (bevel && cross < 0.0f)
+		if (bevel && cross < 0.0f) // If the corner is convex and an acute enough angle, generate a bevel.
 		{
-			if (n+2 >= maxOutVerts)
+			if (numOutVerts + 2 >= maxOutVerts)
+			{
 				return 0;
-			float d = (1.0f - (dx0*dx1 + dy0*dy1))*0.5f;
-			outVerts[n*3+0] = vb[0] + (-dlx0+dx0*d)*offset;
-			outVerts[n*3+1] = vb[1];
-			outVerts[n*3+2] = vb[2] + (-dly0+dy0*d)*offset;
-			n++;
-			outVerts[n*3+0] = vb[0] + (-dlx1-dx1*d)*offset;
-			outVerts[n*3+1] = vb[1];
-			outVerts[n*3+2] = vb[2] + (-dly1-dy1*d)*offset;
-			n++;
+			}
+
+            // Generate two bevel vertices at a distances from B proportional to the angle between the two segments.
+            // Move each bevel vertex out proportional to the given offset.
+			float d = (1.0f - (prevSegmentDir[0] * currSegmentDir[0] + prevSegmentDir[2] * currSegmentDir[2])) * 0.5f;
+
+			outVerts[numOutVerts * 3 + 0] = vertB[0] + (-prevSegmentNormX + prevSegmentDir[0] * d) * offset;
+			outVerts[numOutVerts * 3 + 1] = vertB[1];
+			outVerts[numOutVerts * 3 + 2] = vertB[2] + (-prevSegmentNormZ + prevSegmentDir[2] * d) * offset;
+			numOutVerts++;
+
+			outVerts[numOutVerts * 3 + 0] = vertB[0] + (-currSegmentNormX - currSegmentDir[0] * d) * offset;
+			outVerts[numOutVerts * 3 + 1] = vertB[1];
+			outVerts[numOutVerts * 3 + 2] = vertB[2] + (-currSegmentNormZ - currSegmentDir[2] * d) * offset;
+			numOutVerts++;
 		}
 		else
 		{
-			if (n+1 >= maxOutVerts)
+			if (numOutVerts + 1 >= maxOutVerts)
+			{
 				return 0;
-			outVerts[n*3+0] = vb[0] - dmx*offset;
-			outVerts[n*3+1] = vb[1];
-			outVerts[n*3+2] = vb[2] - dmy*offset;
-			n++;
+			}
+
+            // Move B along the miter direction by the specified offset.
+			outVerts[numOutVerts * 3 + 0] = vertB[0] - cornerMiterX * offset;
+			outVerts[numOutVerts * 3 + 1] = vertB[1];
+			outVerts[numOutVerts * 3 + 2] = vertB[2] - cornerMiterZ * offset;
```

Comment:
@memononen Digging more into this, I'm still a bit confused why this is _subtracting_ the scaled miter offset from vertB?  Seems like that implies either positive values for `offset` should _shrink_ the polygon instead of expand it, _or_ the bisector here should be actually pointing _into_ the polygon?

e.g.
(I know the code would bevel a 90 degree corner like this.  Just using these numbers to make the math easy.)

```
            C
B *---------*
  |
  |
  |
  |
A *


A = <0, 0, 0>
B = <0, 0, 10>
C = <10, 0, 10>

// Normalized segment directions
prevSegmentDir = normalize(B - A) = <0, 1>
currSegmentDir = normalize(C - B) = <1, 0>

// Segment normals pointing out of the poly (ccw perp vector assuming cw polygon winding order)
prevSegmentNorm = <-1, 0>
currSegmentNorm = <0, 1>

cornerMiter = (prevSegmentNorm + currSegmentNorm) / 2 = <-.5, .5>
cornerMiterSqMag = .5

// The final scaled miter vector (bisector) points out of the polygon
cornerMiter = cornerMiter / cornerMiterSqMag = <-1, 1>

// given an offset value of 2, B is offset correctly but in the wrong direction?
outVert = B - cornerMiter * offset = <0, 0, 10> - <-1, 0, 1> * 2 = <2, 0, 8>

// I'd expect the final result here to be <-2, 0, 12> instead
```
I suspect I'm missing something critical here, or maybe I'm getting turned around with coordinate system handedness?

Reply:
The above is kinda confusing.... when I use 2-component vectors `<1, 2>` it's really an XZ plane vector so think of it as `<1, 0, 2>`

Reply:
The code is the same as in https://github.com/recastnavigation/recastnavigation/blob/main/RecastDemo/Source/imguiRenderGL.cpp#L61 (which is recycled from another rendering lib, which eventually turned into nanovg),
but adjusted to what ever winding the polygon happened to have.